src/settings.rs

```diff
@@ -1,34 +1,121 @@
+        let level = if let Some(ref level) = selinux_options_external.level {
+            Some(SELinuxLevel::new(level.clone())?)
+        } else {
+            None
+        };
+        Ok(SELinuxOptions {
+            user: selinux_options_external.user.clone(),
+            role: selinux_options_external.role.clone(),
+            type_: selinux_options_external.type_.clone(),
+            level,
+        })
+    }
+}
+
+impl From<SELinuxOptions> for apicore::SELinuxOptions {
+    fn from(selinux_options: SELinuxOptions) -> apicore::SELinuxOptions {
+        apicore::SELinuxOptions {
+            level: selinux_options.level.map(|level| level.level),
+            role: selinux_options.role,
+            type_: selinux_options.type_,
+            user: selinux_options.user,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub(crate) struct SELinuxLevel {
+    level: String,
+    sensitivity: String,
+    categories: Vec<String>,
+}
+
+impl PartialEq for SELinuxLevel {
+    fn eq(&self, selinux_level: &SELinuxLevel) -> bool {
+        if self.level == selinux_level.level {
+            return true;
+        }
+        self.sensitivity == selinux_level.sensitivity && self.categories == selinux_level.categories
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl SELinuxLevel {
+    pub(crate) fn new(level: String) -> Result<SELinuxLevel> {
+        let mut splitted_level = level.split(':');
+        if splitted_level.clone().count() != 2 {
+            return Ok(SELinuxLevel {
+                level,
+                ..SELinuxLevel::default()
+            });
+        }
+        let (sensitivity, categories) = (
+            splitted_level.next().unwrap(),
+            splitted_level.next().unwrap(),
+        );
+        let mut splitted_categories: Vec<String> = categories
+            .split(',')
+            .into_iter()
+            .map(String::from)
+            .collect();
+        splitted_categories.sort();
+        Ok(SELinuxLevel {
+            level: level.clone(),
+            sensitivity: sensitivity.to_string(),
+            categories: splitted_categories,
+        })
+    }
+}
 
-    use kubewarden_policy_sdk::settings::Validatable;
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(tag = "rule")]
+pub(crate) enum Rule {
+    MustRunAs(SELinuxOptions),
+    RunAsAny,
+}
 
-    #[test]
-    fn validate_settings() -> Result<(), ()> {
-        let settings = Settings {};
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(default)]
+pub(crate) struct Settings {
+    #[serde(flatten)]
+    pub rule: Rule,
+}
+
+impl Default for Settings {
+    fn default() -> Settings {
+        Settings {
+            rule: Rule::RunAsAny,
+        }
+    }
+}
 
-        assert!(settings.validate().is_ok());
+impl kubewarden::settings::Validatable for Settings {
+    fn validate(&self) -> Result<(), String> {
         Ok(())
     }
 }
```

Comment:
Are the settings always valid?

What happens if:

* `rule` is `RunAsAny` but the user then specifies things like `user`, `role`,...
* when the rule is `MustRunAs`, shouldn't we ensure all the `user`, `role`,... attributes are specified?

Reply:
First part should be covered by serde as I understand it. I have opened https://github.com/serde-rs/serde/issues/2123 to clarify.

Second part is implemented in commit https://github.com/kubewarden/psp-selinux/pull/1/commits/9a78fce712ff7fb8c9ebb872fa6d2708ef7c14b3.